agent/lib/kontena/service_pods/lifecycle_hook_manager.rb

```diff
@@ -0,0 +1,163 @@
+
+        pre_start_hooks.each do |hook|
+          begin
+            info "running pre_start hook: #{hook['cmd']}"
+            service_config = config_container(service_pod.dup, hook['cmd'])
+            service_config['name'] = "#{service_config['name']}-#{SecureRandom.urlsafe_base64(5)}"
+            service_container = create_container(service_config)
+            log_hook_output(service_container.id, "running pre_start hook: #{hook['cmd']}", 'stdout')
+            service_container.tap(&:start).attach
+            if service_container.state['ExitCode'] != 0
+              raise "Failed to execute pre_start hook: #{hook['cmd']}, exit code: #{service_container.state['ExitCode']}"
+            end
+          ensure
+            cleanup_container(service_container) if service_container
+          end
+        end
+        true
+      rescue => exc
+        log_service_pod_event("service:create_instance", exc.message, Logger::ERROR)
+        raise exc
+      end
+
+      # @param [Docker::Container] service_container
+      # @return [Boolean]
+      def on_post_start(service_container)
+        hooks_for('post_start').each do |hook|
+          info "running post_start hook: #{hook['cmd']}"
+          command = build_cmd(hook['cmd'])
+          log_hook_output(service_container.id, "running post_start hook: #{hook['cmd']}", 'stdout')
+          _, _, exit_code = service_container.exec(command) { |stream, chunk|
+            log_hook_output(service_container.id, chunk, stream)
+          }
+          if exit_code != 0
+            raise "Failed to execute post_start hook: #{hook['cmd']} (exit code: #{exit_code}"
+          end
+        end
+        true
+      rescue => exc
+        log_service_pod_event("service:create_instance", exc.message, Logger::ERROR)
+        error exc.message
+        false
+      end
+
+      # @param [Docker::Container] service_container
+      # @return [Boolean]
+      def on_pre_stop(service_container)
+        hooks_for('pre_stop').each do |hook|
+          info "running pre_stop hook: #{hook['cmd']}"
+          command = build_cmd(hook['cmd'])
+          log_hook_output(service_container.id, "running pre_stop hook: #{hook['cmd']}", 'stdout')
+          _, _, exit_code = service_container.exec(command) { |stream, chunk|
+            log_hook_output(service_container.id, chunk, stream)
+          }
+          if exit_code != 0
+            raise "Failed to execute pre_stop hook: #{hook['cmd']} (exit code: #{exit_code}"
+          end
+        end
+        true
+      rescue => exc
+        log_service_pod_event("service:create_instance", exc.message, Logger::ERROR)
+        error exc.message
+        false
+      end
+
+      # @param [String] type
+      # @return [Array<Hash>]
+      def hooks_for(type)
+        hooks = service_pod.hooks.select { |h|
+          h['type'] == type && !cached_oneshot_hook?(h)
+        }
+        hooks.each { |h|
+          mark_oneshot_hook(h) if !cached_oneshot_hook?(h)
```

Comment:
This is also reporting non-oneshot hooks now:

```
D, [2017-09-20T13:29:26.844311 #1] DEBUG -- Kontena::ServicePods::LifecycleHookManager: mark oneshot hook: {"id"=>"59c26d311c7aaf000fe67683", "type"=>"post_start", "cmd"=>"date && echo \"first hook (non-oneshot)\"", "oneshot"=>false}
E, [2017-09-20T13:29:26.922872 #1] ERROR -- Kontena::ServicePods::LifecycleHookManager: RuntimeError: Hook not found: {"id"=>"59c26d311c7aaf000fe67683", "type"=>"post_start", "cmd"=>"date && echo \"first hook (non-oneshot)\"", "oneshot"=>false}
```

```
E, [2017-09-20T13:29:26.902442 #12] ERROR -- RpcServer: RuntimeError: Hook not found: {"id"=>"59c26d311c7aaf000fe67683", "type"=>"post_start", "cmd"=>"date && echo \"first hook (non-oneshot)\"", "oneshot"=>false}
D, [2017-09-20T13:29:26.902574 #12] DEBUG -- RpcServer: /app/app/services/rpc/node_service_pod_handler.rb:81:in `mark_oneshot_hook'
/app/app/services/rpc_server.rb:63:in `handle_request'
/app/app/services/rpc_server.rb:47:in `process!'
/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/calls.rb:28:in `public_send'
/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/calls.rb:28:in `dispatch'
/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/call/async.rb:7:in `dispatch'
/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/cell.rb:50:in `block in dispatch'
/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/cell.rb:76:in `block in task'
/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/actor.rb:339:in `block in task'
/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/task.rb:44:in `block in initialize'
/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/task/fibered.rb:14:in `block in create'
```

Reply:
Fixed.